### PR TITLE
systemctl: reduce variable scope and simplify

### PR DESCRIPTION
**Pull request by dtardon**

### CODE REVIEW

File: src/systemctl/systemctl-list-units.c [hunk 1]
```diff
@@ -403,19 +402,18 @@ static int output_sockets_list(struct socket_info *socket_infos, unsigned cs) {
                                 return table_log_add_error(r);
 
                 }
-
```
Comment (yuwata) [inline src/systemctl/systemctl-list-units.c:406]: If you touch here, then could you also update `if (cs)` -> `if (cs > 0)` in the above?

File: src/systemctl/systemctl-list-units.c [hunk 2]
```diff
@@ -403,19 +402,18 @@ static int output_sockets_list(struct socket_info *socket_infos, unsigned cs) {
...
 
         r = output_table(table);
         if (r < 0)
                 return r;
 
         if (arg_legend != 0) {
+                const char *on, *off;
+
+                on = cs > 0 ? ansi_highlight() : ansi_highlight_red();
+                off = ansi_normal();
```
Comment (yuwata) [inline src/systemctl/systemctl-list-units.c]: These are used only once. I think it is ok to inline them in the `printf()` below, if you'd like.

**Review by yuwata — APPROVED** [review]
LGTM. Several minor suggestions.

File: src/systemctl/systemctl-list-units.c [hunk 3]
```diff
@@ -403,19 +402,18 @@ static int output_sockets_list(struct socket_info *socket_infos, unsigned cs) {
                                 return table_log_add_error(r);
 
                 }
-
-                on = ansi_highlight();
-                off = ansi_normal();
-        } else {
-                on = ansi_highlight_red();
-                off = ansi_normal();
         }
```
Comment (yuwata) [inline src/systemctl/systemctl-list-units.c:410]: You can drop this `}` now.

File: src/systemctl/systemctl-list-units.c [hunk 4]
```diff
@@ -403,19 +402,18 @@ static int output_sockets_list(struct socket_info *socket_infos, unsigned cs) {
                                 return table_log_add_error(r);
```
Comment (yuwata) [inline src/systemctl/systemctl-list-units.c:383]: Please also drop this empty line.

File: src/systemctl/systemctl-list-units.c (hunk 1 above)
Comment (yuwata) [reply src/systemctl/systemctl-list-units.c:406]: I noticed that the condition `if (cs)` can be dropped now.

Comment (dtardon) [reply src/systemctl/systemctl-list-units.c:406]: Done.

File: src/systemctl/systemctl-list-units.c (hunk 3 above)
Comment (dtardon) [reply src/systemctl/systemctl-list-units.c:410]: Done.

File: src/systemctl/systemctl-list-units.c (hunk 4 above)
Comment (dtardon) [reply src/systemctl/systemctl-list-units.c:383]: Done.

File: src/systemctl/systemctl-list-units.c (hunk 2 above)
Comment (dtardon) [reply src/systemctl/systemctl-list-units.c]: Okay, let's try that.

File: src/systemctl/systemctl-list-units.c [hunk 5]
```diff
@@ -405,14 +415,8 @@ static int output_sockets_list(struct socket_info *socket_infos, unsigned cs) {
...
 
-        if (arg_legend != 0) {
-                printf("\n%s%u sockets listed.%s\n",
-                       cs > 0 ? ansi_highlight() : ansi_highlight_red(),
-                       cs,
-                       ansi_normal());
-                if (!arg_all)
-                        printf("Pass --all to see loaded but inactive sockets, too.\n");
-        }
+        if (arg_legend != 0)
```
Comment (medhefgo) [inline src/systemctl/systemctl-list-units.c:416]: Might as well move this into the function too…

Comment (dtardon) [reply src/systemctl/systemctl-list-units.c:416]: I wanted to keep it explicit that the output of legend is optional. But I don't care much either way...

File: src/systemctl/systemctl-list-units.c [hunk 6]
```diff
@@ -641,20 +635,15 @@ static int output_timers_list(struct timer_info *timer_infos, unsigned n) {
...
         r = output_table(table);
         if (r < 0)
                 return r;
 
         if (arg_legend != 0) {
-                printf("\n%s%u timers listed.%s\n", on, n, off);
+                printf("\n%s%u timers listed.%s\n",
+                       n > 0 ? ansi_highlight() : ansi_highlight_red(),
+                       n,
+                       ansi_normal());
```
Comment (poettering) [inline src/systemctl/systemctl-list-units.c]: I am not sure these are really an improvement, no? Sure, it's a bit shorter, but I kinda liked the fact the coloring was separate from the actual string we output here. Now it's a more complex expression where it's harder to grok what is what...

File: src/systemctl/systemctl-list-units.c [hunk 7]
```diff
@@ -369,38 +369,36 @@ static int output_sockets_list(struct socket_info *socket_infos, unsigned cs) {
 
         (void) table_set_empty_string(table, "-");
 
-        if (cs) {
-                for (struct socket_info *s = socket_infos; s < socket_infos + cs; s++) {
```
Comment (poettering) [inline src/systemctl/systemctl-list-units.c:374]: I have the suspicion some of the static checkers are not going to like this, given that socket_infos might be NULL here. i.e. the static checkers really don't like it if one does maths on NULL pointers, even if the value you add to them is 0 itself...

Hence, change the second part of the for expression from this:

```c
…; s < socket_infos + cs; …
```

to

```c
…; socket_infos && (s < socket_infos + cs); …
```

then the static checkers shouldn't mind

**Comment by poettering** [comment]
looks mostly good, but the first commit i don#t find convincing, and the static checking thing should be fixed, before the static checker complains, even though it's not a real issue, just the checkers are too dumb.

File: src/systemctl/systemctl-list-units.c (hunk 6 above)
Comment (dtardon) [reply src/systemctl/systemctl-list-units.c]: Well, my original patch just reduced scope of the two vars and shortened their definition. The inlining was @yuwata's suggestion. (And the extraction of the legend output to a separate function was an afterthought too.) I'm fine either way.

**Comment by dtardon** [comment]
Btw, I've just realized it might make sense to merge this to #24033, because that could then use `output_legend()` too.

File: src/systemctl/systemctl-list-units.c [hunk 8]
```diff
@@ -197,13 +197,11 @@ static int output_units_list(const UnitInfo *unit_infos, unsigned c) {
...
                                 puts("JOB    = Pending job for the unit.\n");
-                        on = ansi_highlight();
-                        off = ansi_normal();
-                } else {
-                        on = ansi_highlight_red();
-                        off = ansi_normal();
                 }
 
+                on = records > 0 ? ansi_highlight() : ansi_highlight_red();
+                off = ansi_normal();
```
Comment (poettering) [inline src/systemctl/systemctl-list-units.c:203]: I think the first commit should change the code like this, too. I think this is the most readable + relatively short way to spell this out.

**Comment by poettering** [comment]
btw, while reviewing this, i noticed some issues that this PR fixes: #24105. Merging that will cause conflicts in this PR, and vice versa.

It might make sense to pull that PR into yours?

Either way, this PR should make similar fixes, i.e. in the second to last commit, the `n_items` parameter should also be switched from `unsigned` to `size_t`.

**Comment by dtardon** [comment]
I've merged this to #24033 -> closing.

**Pull request closed without merging**